### PR TITLE
Flow Graph Path Mapping

### PR DESCRIPTION
**Pull request by carolhmj**
# Changes
 - On the SetPropertyBlock, add a new parameter named `subString`, which represents a string that will be substituted by a node value if encountered between `{}`. For example, if the subString is `nodeIndex`, and the path is `/node/{nodeIndex}/`, when this node is activated, it will look into the nodeIndex socket, and substitute its value (which could be 0, then the path would be `/node/0`). This allows dynamic paths that are resolved based on the current state of the scene.
   - This change will be needed because of how the glTF Interactivity uses paths. In some cases, paths are defined as `/node/{nodeIndex}/property`, and we want these paths to be resolved at runtime.
-  The SetProperty node was changed to use a configuration object instead of sockets. The configuration object contains a string path (as defined in the context variables), and a string property, such as translation, rotation, etc.
- I also moved the default serialization and parsing functions to a separate module, and added handling of vector values.
- I also removed the variableDefinitions value, as it wasn't being used and made things more complex.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14481/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14481/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14481/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14481/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14481/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14481/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14481/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14481/merge#BCU1XR#0

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 1]
```diff
@@ -96,6 +61,11 @@ export class FlowGraphContext {
      */
     private _executionId = 0;
 
+    /**
+     * A mapping of paths to target objects
+     */
+    pathMap: Map<string, any> = new Map();
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:67]: nit - a simple key-value object will suffice. Unless you are planning on using the maps' functionality?

**Review by RaananW — COMMENTED** [review]
I think it can be abstracted a bit more, to support future changes.
There should be a path, and a configurable map of replacement values. it should then take the value from the context, based on the map key, instead of fixing the nodeIndex as a property. 
GetTargetFromPath can be a static member of FlowGraph (or the context). This was it can be used on other future blocks as well.

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 2]
```diff
@@ -274,6 +257,7 @@ export class FlowGraphContext {
         this._connectionValues.forEach((value, key) => {
             valueSerializationFunction(key, value, serializationObject._connectionValues);
         });
+        serializationObject.pathMap = {};
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:260]: probably not needed (was removed?)

File: packages/dev/core/test/unit/FlowGraph/flowGraphSerialization.test.ts [hunk 3]
```diff
@@ -128,6 +128,8 @@ describe("Flow Graph Serialization", () => {
         context.setVariable("test3", new Vector3(1, 2, 3));
         context.setVariable("test4", mesh);
 
+        context.pathMap.set("testPath", mesh);
```
Comment (RaananW) [inline packages/dev/core/test/unit/FlowGraph/flowGraphSerialization.test.ts:131]: this was removed, so can be removed from here as well

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 4]
```diff
@@ -259,6 +224,24 @@ export class FlowGraphContext {
...
+
+    /** @internal */
+    public _getTargetFromPath(path: string, subString: string, block: FlowGraphBlock) {
+        let finalPath = path;
+        if (subString && path.indexOf(this._getEnclosedSubstring(subString)) !== -1) {
+            const nodeSub = block.getDataInput(subString);
+            if (!nodeSub) {
+                throw new Error("Invalid substitution input");
+            }
+            const nodeIndex = Math.floor(nodeSub.getValue(this));
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:239]: was discussed - expected to be a number (and should be validated)
Also renaming to index might be better naming

**Review by RaananW — COMMENTED** [review]
Looks great! just a few nitty comments.

**Review by RaananW — APPROVED** [review]

**Pull request merged**